(DOCSP-18849): Add build and hosting to web tutorial (#1381)
mongodb/docs-realm: source/tutorial/web-graphql.txt

.. code-block:: shell
   
   npx serve -s build

You will see a new browser window pop up running the app on ``localhost:5000``. The app should 
look and behave the same as the development version you made in the previous step. 

You can find the production bundle's code with your repository in a newly generated folder ``build``.

.. _deploy-web-app-realm-ui:

Deploy Using the Realm UI
~~~~~~~~~~~~~~~~~~~~~~~~~

First, navigate to the **Hosting** section of the Realm UI. You can find it on the side menu under **Manage**. 
Once you're on the Hosting page, click the **Enable Hosting** button. 

Now that you have hosting enabled, replace the default ``index.html`` with the production 
bundle you built in the last step. Delete the ``index.html`` file by clicking the check 
box next to the file and then **Actions > Delete**. 

To add your app, open a file browser on your computer, and navigate to the ``build`` 
folder in your app's root folder. Select all the ``build`` folder's contents, and 
drag and drop it over to the Realm Hosting UI's Files page. 

Once the files finish uploading, click the **Review & Deploy** button at the top of the page. 

> A dialog will open with you changes. Review the changes and click the **Deploy** button

Reviewer: ```suggestion
A dialog will open with your changes. Review the changes and click the **Deploy** button
```

at the bottom of the dialogue to start the deployment of your app.

> Realm then deploys your site to ``<Your-App-Id>.mongodbstitch.com``. 

Reviewer: ```suggestion
Realm deploys your site to ``<Your-App-Id>.mongodbstitch.com``. 
```

Deployment can take up to a few minutes. You can track the status on the Hosting page. 

Once deployment finishes, navigate to your app's new URL, ``<Your-App-Id>.mongodbstitch.com``. 
Here you'll find the app with the same backend and data that you were working with 

> before, now live on the internet!

Reviewer: ```suggestion
before, now live on the Internet!
```

For more detailed information about hosting your web app with Realm, refer to our documentation 
on :ref:`hosting a single page application <host-a-single-page-application>`.

.. _deploy-web-app-realm-cli:

.. note:: Deploy Using the Realm CLI

   You can also deploy the app frontend with the Realm CLI. 
   Before you're able to deploy the frontend from the Realm CLI, you must first 
   enable hosting in the Realm UI, as discussed at the beginning of the 
   :ref:`Deploy Using the Realm UI <deploy-web-app-realm-ui>` section. 

>    Once you've enabled hosting in the UI, you need to have the `realm-tutorial-backend <https://github.com/mongodb-university/realm-tutorial-backend>`_ 

Reviewer: Why do you need to enable hosting in the UI (or visit the UI at all) if you're about to set it up in the config below?

   repository handy, which you used in :ref:`the backend setup <tutorial-task-tracker-create-realm-app>`.
   
   In the **backend** directory, create the folder ``hosting/files``. This is where 
   you will place your assets for Realm Hosting. 

   Also create a config file for the app. This should be located at ``hosting/config.json``, 
   and have the following contents: 

   .. code-block:: json
      
      {
         "enabled": true,
         "default_response_code": 200,
         "default_error_path": "/index.html",
      }

   Copy the contents of the ``build`` folder from the realm-tutorial-web directory  
   to the ``hosting/files`` folder in the backend. You can do this with the command:

   .. code:: shell
      
      cp -a path/to/realm-tutorial-web/build/. path/to/realm-tutorial-backend/hosting/files

>    Now you're ready to deploy the app using the Realm CLI. To deploy, simply run the 

Reviewer: ```suggestion
   Now you're ready to deploy the app using the Realm CLI. To deploy, run the 
```

- Navigate to *second@example.com*'s project
- Collaborate by adding, updating, and removing some new tasks

> I. Build the App

Reviewer: "Build" might be ambiguous. Should we say "Deploy"? I think this step and the next could be merged, as building a production bundle isn't that special to Realm.

----------------

Now that you've validated that the app is working, it's time to get the app ready for deployment. 
In this section, you're going to create a production build with `Create React App's build script <https://create-react-app.dev/docs/production-build/>`__.

To create your production bundle, run the command: 

.. code-block:: shell
   
   npm run build

Once the build finishes, validate it's working by running:

.. code-block:: shell
   
   npx serve -s build

You will see a new browser window pop up running the app on ``localhost:5000``. The app should 
look and behave the same as the development version you made in the previous step. 

You can find the production bundle's code with your repository in a newly generated folder ``build``.

J. Deploy the App with Realm
----------------------------

> With the production bundle ready, it's time to deploy the app on the web. 

Reviewer: This sentence can be removed if you merge the two steps as suggested earlier.

> You're going to deploy the app using :ref:`Realm Static Hosting <static-hosting>` via the Realm UI.

Reviewer: You can simplify by using imperative mood:
```suggestion
Deploy the app using :ref:`Realm Static Hosting <static-hosting>` via the Realm UI.
```

.. note:: Deploying App with Realm CLI
   You can also :ref:`deploy your app to static hosting using the Realm CLI <host-a-single-page-application>`.

>    However, in this tutorial we're just going to explore deployment using the Realm UI. 

Reviewer: Yes, but it muddies the point a bit. I would just say:
```suggestion
   However, in this tutorial we're just going to deploy using the Realm UI. 
```

It might be even better just to say, "This tutorial step uses the Realm UI to deploy."

First, navigate to the **Hosting** section of the Realm UI. You can find it on the side menu under **Manage**. 

> Once you're on the Hosting page, click the **Enable Hosting** button. 

Reviewer: Could this step be done as part of the imported backend?

Now that you have hosting enabled, replace the default ``index.html`` with the production 
bundle you built in the last step. Delete the ``index.html`` file by clicking the check 
box next to the file and then **Actions > Delete**. 

To add your app, first open a file browser on your computer, such as Finder on Mac or 
File Explorer on Windows. Navigate to the ``build`` folder in your app's root folder. 
Select all the ``build`` folder's contents, and drag it over to the Realm Hosting UI's 

> Files page. Release the mouse to trigger uploading the assets. 

Reviewer: This is way too detailed. Our readers should know how drag and drop works.

Once the files finish uploading, click the **Review & Deploy** button at the top of the page. 
A dialog will open with you changes. Review the changes and click the **Deploy** button
at the bottom of the dialogue to start the deployment of your app.

> Realm then will deploy your site to ``<Your-App-Id>.mongodbstitch.com``. 

Reviewer: For clarity/brevity's sake (even if it's not 100% temporally suited to your intention), avoid future tense:
```suggestion
Realm deploys your site to ``<Your-App-Id>.mongodbstitch.com``. 
```

Deployment can take up to a few minutes. You can track the status on the Hosting page. 

> Once deployment finishes, navigate to your app's new domain. Here you'll find the 

Reviewer: "New domain" could be clarified. I think you mean app URL. You might want to mention where that can be found again.
